crates/ecs/src/systems/command_buffers.rs

```diff
@@ -0,0 +1,742 @@
+        let entity0 = app.create_entity().unwrap();
+        app.add_component(entity0, D).unwrap();
+        app.add_component(entity0, E).unwrap();
+        app.add_component(entity0, F).unwrap();
+        let entity1 = app.create_entity().unwrap();
+        app.add_component(entity1, D).unwrap();
+        app.add_component(entity1, E).unwrap();
+        let entity2 = app.create_entity().unwrap();
+        app.add_component(entity2, D).unwrap();
+        (app, entity0, entity1, entity2)
+    }
+
+    fn remaining_entities_in_world(world: &World) -> Vec<Entity> {
+        world.entities.iter().cloned().flatten().collect()
+    }
+
+    #[test]
+    fn system_can_remove_entities_until_next_tick() {
+        let removing_system = |entity: Entity, commands: Commands| {
+            commands.remove(entity);
+        };
+
+        let (app, _, _, _) = set_up_app_with_systems_and_entities([removing_system]);
+
+        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();
+        runner.tick().unwrap();
+        runner.playback_commands().unwrap();
+
+        let remaining_entities = remaining_entities_in_world(&runner.world);
+        assert!(
+            remaining_entities.is_empty(),
+            "all entities should be removed, but these remain: {:?}",
+            remaining_entities
+        )
+    }
+
+    #[test]
+    fn system_removing_same_entity_second_time_is_a_noop() {
+        let twice_removal_system = |entity: Entity, commands: Commands| {
+            commands.remove(entity);
+            commands.remove(entity);
+        };
+
+        let (app, _, _, _) = set_up_app_with_systems_and_entities([twice_removal_system]);
+
+        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();
+        runner.tick().unwrap();
+        runner.playback_commands().unwrap();
+
+        let remaining_entities = remaining_entities_in_world(&runner.world);
+        assert!(
+            remaining_entities.is_empty(),
+            "all entities should be removed, but these remain: {:?}",
+            remaining_entities
+        )
+    }
+
+    #[test]
+    fn two_systems_removing_same_entity_removes_that_entity_once() {
+        let removing_system0 = |entity: Entity, commands: Commands| {
+            commands.remove(entity);
+        };
+        let removing_system1 = |entity: Entity, commands: Commands| {
+            commands.remove(entity);
+        };
+
+        let (app, _, _, _) =
+            set_up_app_with_systems_and_entities([removing_system0, removing_system1]);
+
+        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();
+        runner.tick().unwrap();
+        runner.playback_commands().unwrap();
+
+        let remaining_entities = remaining_entities_in_world(&runner.world);
+        assert!(
+            remaining_entities.is_empty(),
+            "all entities should be removed, but these remain: {:?}",
+            remaining_entities
+        )
+    }
+
```

Comment:
I found an issue where if you have **one system adding components** and **one removing the same entities**, it results in an unexpected crash, caused by `.expect("Entity should have a component index tied to it since it is already established to be existing within the archetype")` on line 477 in archetypes.rs.

Reply:
If you do the same thing but with `commands.remove_component` it works correctly, so it seems reasonable that both cases should be allowed.

Reply:
All seems to work! Close this and MERGE !